internal/sqldb/writer.go

```diff
@@ -113,56 +130,122 @@ func Write(sqlClient *sql.DB, resourceMetadata *resource.Metadata) error {
+		if len(parts) == 0 {
+			return "", "", false
+		}
+		bucketName := parts[0]
+		objectPrefix := ""
+		if len(parts) == 2 {
+			objectPrefix = parts[1]
+			if !strings.HasPrefix(objectPrefix, "/") {
+				objectPrefix += "/"
+			}
+		}
+		log.Printf("bucket: %s, prefix: %s", bucketName, objectPrefix)
+		return bucketName, objectPrefix, true
 	}
+	return "", "", false
+}
 
-	var res []string
-	for _, file := range files {
-		if fName := file.Name(); strings.HasSuffix(fName, ".csv") {
-			res = append(res, file.Name())
+// Get csv file handle.
+// Make sure to close the file returned from this function.
+func listCSVFiles(dir string) ([]*csvHandle, error) {
+	var res []*csvHandle
+	if bucketName, objectPrefix, ok := parseGCSPath(dir); ok {
+		// Read from GCS
+		ctx := context.Background()
+		client, err := storage.NewClient(ctx)
+		if err != nil {
+			return nil, err
+		}
+		defer client.Close()
+		bucket := client.Bucket(bucketName)
+		query := &storage.Query{
+			Prefix: objectPrefix,
+		}
+		it := bucket.Objects(ctx, query)
+		for {
+			objAttrs, err := it.Next()
+			if err == iterator.Done {
+				break
+			}
+			if err != nil {
+				return nil, err
+			}
+			// Check if it's a CSV file
+			if strings.HasSuffix(objAttrs.Name, ".csv") {
+				rc, err := bucket.Object(objAttrs.Name).NewReader(ctx)
+				if err != nil {
+					return nil, err
+				}
+				log.Printf("Added csv: %s", objAttrs.Name)
+				res = append(
+					res,
+					&csvHandle{
+						f:     rc,
+						name:  objAttrs.Name,
+						close: func() { rc.Close() },
+					},
+				)
+			}
+		}
+	} else {
+		// Read from local files
+		files, err := os.ReadDir(dir)
+		if err != nil {
+			return nil, err
+		}
+		for _, file := range files {
+			if fName := file.Name(); strings.HasSuffix(fName, ".csv") {
+				f, err := os.Open(filepath.Join(dir, fName))
+				if err != nil {
+					return nil, err
+				}
+				res = append(
+					res,
+					&csvHandle{
+						f:     f,
+						name:  fName,
+						close: func() { f.Close() },
+					},
+				)
+			}
 		}
 	}
-
 	return res, nil
 }
 
 func processCSVFile(
 	medatata *resource.Metadata,
-	fileDir string,
-	csvFile string,
+	ch *csvHandle,
 	provID string,
 ) (
 	[]*observation,
 	[]string, // A list of variables.
 	error,
 ) {
-	// Read the CSV file.
-	f, err := os.Open(filepath.Join(fileDir, csvFile))
-	if err != nil {
-		return nil, nil, err
-	}
-	defer f.Close()
-	records, err := csv.NewReader(f).ReadAll()
+	records, err := csv.NewReader(ch.f).ReadAll()
 	if err != nil {
 		return nil, nil, err
 	}
 	numRecords := len(records)
 	if numRecords < 2 {
 		return nil, nil, status.Errorf(codes.FailedPrecondition,
-			"Empty CSV file %s", csvFile)
+			"Empty CSV file %s", provID)
 	}
 
 	// Load header.
 	header := records[0]
 	if len(header) < 3 {
 		return nil, nil, status.Errorf(codes.FailedPrecondition,
-			"Less than 3 columns in CSV file %s", csvFile)
+			"Less than 3 columns in CSV file %s", provID)
 	}
 	numColumns := len(header)
 
```

Comment:
Based on our earlier discussion, the resolving logic will be removed and mixer will only load pre-resolved entities, correct?

Reply:
I will extend mixer to take resolved entities, but it should still keep the capability to take unresolved csv.